File: backend/app/schemas/admin_datasets.py
import uuid
from datetime import date, datetime
from typing import Literal

from pydantic import BaseModel, Field

from app.schemas.datasets import DatasetFilePublic


class DatasetCreateMinimal(BaseModel):
    """Minimal dataset-shell creation, sufficient to attach files to for
    Phase 2 ingestion testing. The full DatasetModal-equivalent create/edit
    form (category assignment, platforms, processing levels, publish status,
    etc.) is Phase 8 scope — this endpoint intentionally does not attempt to
    replicate that yet."""

    title: str = Field(min_length=1, max_length=500)
    description: str | None = None
    code: str | None = Field(default=None, max_length=50)


class DatasetMinimalPublic(BaseModel):
    id: uuid.UUID
    code: str
    title: str
    description: str | None
    status: str

    model_config = {"from_attributes": True}


class DatasetCreate(BaseModel):
    """Full admin create/edit form payload. All fields beyond title/
    description are optional so this schema also validates the minimal
    Phase 2 payload shape (`{title, description}` or `{title, description,
    code}`) on the same `POST /admin/datasets` route."""

    title: str = Field(min_length=1, max_length=500)
    description: str | None = None
    code: str | None = Field(default=None, max_length=50)
    category_id: uuid.UUID | None = None
    location: str | None = Field(default=None, max_length=255)
    source: str | None = Field(default=None, max_length=255)
    platforms: list[str] = Field(default_factory=list)
    parameters: list[str] = Field(default_factory=list)
    resolution: str | None = Field(default=None, max_length=100)
    license: str | None = Field(default=None, max_length=255)
    processing_levels: list[str] = Field(default_factory=list)
    status: Literal["draft", "published", "archived"] = "draft"


class DatasetUpdate(BaseModel):
    title: str | None = Field(default=None, min_length=1, max_length=500)
    description: str | None = None
    category_id: uuid.UUID | None = None
    location: str | None = Field(default=None, max_length=255)
    source: str | None = Field(default=None, max_length=255)
    platforms: list[str] | None = None
    parameters: list[str] | None = None
    resolution: str | None = Field(default=None, max_length=100)
    license: str | None = Field(default=None, max_length=255)
    processing_levels: list[str] | None = None


class DatasetAdminSummary(BaseModel):
    id: uuid.UUID
    code: str
    title: str
    category_name: str | None
    location: str | None
    record_count: int
    status: str
    updated_at: datetime

    model_config = {"from_attributes": True}


class DatasetAdminDetail(BaseModel):
    id: uuid.UUID
    code: str
    title: str
    description: str | None
    category_id: uuid.UUID | None
    category_name: str | None
    location: str | None
    source: str | None
    platforms: list[str]
    parameters: list[str]
    resolution: str | None
    license: str | None
    processing_levels: list[str]
    formats: list[str]
    status: str
    temporal_start: date | None
    temporal_end: date | None
    record_count: int
    created_at: datetime
    updated_at: datetime
    files: list[DatasetFilePublic]
    active_grant_count: int

    model_config = {"from_attributes": True}


class DatasetPermanentDeleteConfirm(BaseModel):
    confirm: Literal[True]

File: frontend/src/lib/types/admin-datasets.ts
// Mirrors backend/app/schemas/admin_datasets.py exactly — keep these two in sync.

export interface DatasetFilePublic {
  id: string;
  dataset_id: string;
  file_name: string;
  storage_backend: string;
  file_format: string | null;
  file_size_bytes: number | null;
  checksum: string | null;
  temporal_start: string | null;
  temporal_end: string | null;
  version: number;
  file_metadata: Record<string, unknown> | null;
}

export interface UploadStatusResponse {
  id: string;
  dataset_id: string | null;
  dataset_file_id: string | null;
  file_name: string;
  size_bytes: number | null;
  status: string;
  error_message: string | null;
  uploaded_at: string;
  celery_task_id: string | null;
}

export interface DatasetFileUploadResponse {
  upload: UploadStatusResponse;
  dataset_file: DatasetFilePublic;
}

export type DatasetStatus = "draft" | "published" | "archived";

export interface DatasetCreate {
  title: string;
  description?: string | null;
  code?: string | null;
  category_id?: string | null;
  location?: string | null;
  source?: string | null;
  platforms?: string[];
  parameters?: string[];
  resolution?: string | null;
  license?: string | null;
  processing_levels?: string[];
  status?: DatasetStatus;
}

export interface DatasetUpdate {
  title?: string;
  description?: string | null;
  category_id?: string | null;
  location?: string | null;
  source?: string | null;
  platforms?: string[] | null;
  parameters?: string[] | null;
  resolution?: string | null;
  license?: string | null;
  processing_levels?: string[] | null;
}

export interface DatasetAdminSummary {
  id: string;
  code: string;
  title: string;
  category_name: string | null;
  location: string | null;
  record_count: number;
  status: string;
  updated_at: string;
}

export interface DatasetAdminDetail {
  id: string;
  code: string;
  title: string;
  description: string | null;
  category_id: string | null;
  category_name: string | null;
  location: string | null;
  source: string | null;
  platforms: string[];
  parameters: string[];
  resolution: string | null;
  license: string | null;
  processing_levels: string[];
  formats: string[];
  status: string;
  temporal_start: string | null;
  temporal_end: string | null;
  record_count: number;
  created_at: string;
  updated_at: string;
  files: DatasetFilePublic[];
  active_grant_count: number;
}

export interface DatasetPermanentDeleteConfirm {
  confirm: true;
}
